### PR TITLE
added most recent or most done stats option

### PR DESCRIPTION
**Pull request by cfreer**

### CODE REVIEW

**Comment by klavins** [comment]
Interesting. Thanks. A few things I see that might not scale well as this code gets bigger.

1) Split stats() into multiple functions. Generally, we want functions to be fewer than 20 lines long (for now this doesn't need to apply to the switch statement in crush, but eventually we'll replace all that with a argument parsing library and shorten it up).

2) Generally we don't want to use strings as objects with meaning beyond their being strings. So for example, you send  "mr" as an argument. A better way to do this would be to name a function called most_recent() or most_used() and call that. Those functions could call the original stats() function to get the whole list and then return or print a subset of the operation types. That way, each function has a single job.

**Comment by cfreer** [comment]
@klavins made those changes - let me know if I should fix anything else!

File: stats.js [hunk 1]
```diff
@@ -2,21 +2,87 @@ const AQ = require('./gofish/aquarium.js');
...
             console.log("\t" + key + ": " + val);
         }
     }
+}
+
+async function most_recent() {
+    let operation_types = await get_operation_types();
+    let stat_map = {};
+
+    for (let i = 0; i < operation_types.length; i++) {
```
Comment (klavins) [inline stats.js]: This wants to be a method, as it is duplicated code with lines 47-54. I would suggest that you put it in your get_operation_types method. That way, all your methods start with a call to get_operation_types() and you have one place to go to make any changes to that.

File: stats.js [hunk 2]
```diff
@@ -2,21 +2,87 @@ const AQ = require('./gofish/aquarium.js');
...
+        let response = await AQ.get("/operation_types/"+operation_types[i].id+"/stats");
+        let stats = JSON.parse(response.data);
+        console.log(`Processing ${operation_types[i].category}/${operation_types[i].name}`);
+        if (stats.last_run) {
+            stat_map[i] = stats;
```
Comment (klavins) [inline stats.js]: You can actually do something like operation_types[i].stats = JSON.parse(response.data). That way, each operation_type object owns its stats and you don't need more lists.

File: stats.js [hunk 3]
```diff
@@ -2,21 +2,87 @@ const AQ = require('./gofish/aquarium.js');
...
+
+async function get_operation_types () {
+    console.log("Retrieving operation type stats");
+    await connect();
+    let operation_types = await AQ.OperationType.all();
+    return operation_types;
 }
 
-module.exports = stats;
+module.exports = {stats, most_done, most_recent};
```
Comment (klavins) [inline stats.js]: How about defining a class Stats and exporting the Stats class. operation_types would be become this.operation_types so all your methods have access to it. You could put the stats into each this.operation_types[i] or you could have a this.stat_map[i]

Then in the crush main file you would do things like

stats = new Stats(); // the constructor would do what get_operation_types does plus what lines 21-28 do. 

stats.sort_by_last_run();  // or
stats.sort_by_most_done(); // or don't sort

stats.display(stats.all());       // or
stats.display(stats.top(10)); // top(n) would return this.operation_types.slice(0,n)

stats.all() just returns this.operation_types

File: stats.js [hunk 4]
```diff
@@ -2,21 +2,87 @@ const AQ = require('./gofish/aquarium.js');
...
+    for (let i = 0; i < operation_types.length; i++) {
+        let response = await AQ.get("/operation_types/"+operation_types[i].id+"/stats");
+        let stats = JSON.parse(response.data);
+        console.log(`Processing ${operation_types[i].category}/${operation_types[i].name}`);
+        if (stats.last_run) {
+            stat_map[i] = stats;
+        }
+    }
+
+    let sorted_stats = Object.keys(stat_map).map(function(key) {
```
Comment (klavins) [inline stats.js]: If you just add a stats field to operatation_types[i] when you retrieve the stats, you don't need to build this new object before sorting.

**Comment by klavins** [comment]
Added a bunch of comments. Some high level coding practices:

- Avoid duplicate code
- Group methods that operate on the same data into class
- In javascript you can add more data to objects
- Separate concerns, e.g. printing versus sorting (which you did)

File: stats.js [hunk 5]
```diff
@@ -2,21 +2,87 @@ const AQ = require('./gofish/aquarium.js');
...
+}
+
+async function most_done() {
+    let operation_types = await get_operation_types();
+    let stat_map = {};
+
+    for (let i = 0; i < operation_types.length; i++) {
+        let response = await AQ.get("/operation_types/"+operation_types[i].id+"/stats");
+        let stats = JSON.parse(response.data);
+        console.log(`Processing ${operation_types[i].category}/${operation_types[i].name}`);
```
Comment (klavins) [inline stats.js]: If you use something like
process.stdout.clearLine(); 
process.stdout.cursorTo(0); 
process.stdout.write(Math.floor(100*i/operation_types.length) + "%");
you could get a progress percentage. You could put this in a separate method call
show_progress(i,n)

File: stats.js (hunk 1 above)
Comment (cfreer) [reply stats.js]: done!

File: stats.js (hunk 2 above)
Comment (cfreer) [reply stats.js]: done!

File: stats.js (hunk 4 above)
Comment (cfreer) [reply stats.js]: done!

File: stats.js (hunk 5 above)
Comment (cfreer) [reply stats.js]: done!

File: stats.js (hunk 3 above)
Comment (cfreer) [reply stats.js]: done!

**Review by cfreer — COMMENTED** [review]
Finished making suggested changes. Ran into some issues with making Stats a class, since you can't have asynchronous constructors. I found a work-around, but let me know anything else I should change!

**Comment by klavins** [comment]
This looks pretty tight. If you can resolve the conflicts with the changes I made today, then I'll approve the pull request.

**Comment by klavins** [comment]
Oh, and you did the right thing not having an async constructor. Really, all a constructor should do is set up the data structure. It shouldn't do any work.

**Pull request closed without merging**